File: push-fe/src/shared/api/approvals.ts
import { api } from './client'
import { request } from './envelope'
import type { DataEnvelope, ListEnvelope, ListParams } from './envelope'

export type ApprovalKind =
  | 'EVIDENCE_USE'
  | 'DOCUMENT_FINALIZE'
  | 'APPLICATION_SUBMIT'
  | 'CLI_EXECUTE'

export type ApprovalStatus =
  | 'PENDING'
  | 'APPROVED'
  | 'DENIED'
  | 'EXPIRED'
  | 'CONSUMED'

export type Approval = {
  id: string
  revision: number
  kind: ApprovalKind
  applicationId: string
  targetId: string
  targetRevision: number | null
  payloadHash: string
  status: ApprovalStatus
  expiresAt: string | null
  decidedAt: string | null
  consumedAt: string | null
  createdAt: string
  updatedAt: string
}

export type ApprovalSummary = {
  approval: Approval
  targetSummary: {
    title?: string
    workingDirectory?: string
    executable?: string
    arguments?: string[]
    prompt?: string
  } | null
}

export const getApproval = async (id: string): Promise<ApprovalSummary> =>
  request(() => api.get(`approvals/${id}`))

export const listApprovals = async (
  params: ListParams & { applicationId?: string; status?: ApprovalStatus } = {},
): Promise<ListEnvelope<Approval>> =>
  request(() =>
    api.get('approvals', {
      searchParams: {
        ...(params.limit ? { limit: params.limit } : {}),
        ...(params.cursor ? { cursor: params.cursor } : {}),
        ...(params.applicationId ? { applicationId: params.applicationId } : {}),
        ...(params.status ? { status: params.status } : {}),
      },
    }),
  )

export const decideApproval = async (
  id: string,
  body: { expectedRevision: number; decision: 'APPROVED' | 'DENIED' },
): Promise<Approval> => {
  const env = await request<DataEnvelope<Approval>>(() =>
    api.post(`approvals/${id}/decision`, { json: body }),
  )
  return env.data
}

File: push-be/app/presentation/routers/approvals.py
from __future__ import annotations

from fastapi import APIRouter, Request

from app.domain.errors import validation_field
from app.presentation import schemas as s
from app.presentation.deps import (
    bind_json, current_user, data, optional_query_uuid, page_body,
    page_request, param_id,
)

router = APIRouter()

APPROVAL_STATUSES = {"PENDING", "APPROVED", "DENIED", "EXPIRED", "CONSUMED"}


def _deps(request: Request):
    return request.app.state.deps


@router.get("/approvals")
async def list_approvals(request: Request):
    d = _deps(request)
    status = request.query_params.get("status") or None
    if status is not None and status not in APPROVAL_STATUSES:
        raise validation_field("status", "unsupported status")
    p = await d.approvals.list(
        current_user(request).id,
        optional_query_uuid(request, "applicationId"), status,
        page_request(request))
    return page_body(p)


@router.post("/approvals", status_code=201)
async def create_approval(request: Request):
    d = _deps(request)
    req = await bind_json(request, s.CreateApprovalReq)
    a = await d.approvals.create(current_user(request).id, req.kind,
                                 req.application_id, req.target_id,
                                 req.target_revision)
    return data(201, a)


@router.get("/approvals/{id}")
async def get_approval(id: str, request: Request):
    d = _deps(request)
    a, target = await d.approvals.get(current_user(request).id,
                                      param_id(id, "id"))
    return data(200, {"approval": a, "target": target})


@router.post("/approvals/{id}/decision")
async def decide_approval(id: str, request: Request):
    d = _deps(request)
    req = await bind_json(request, s.DecisionReq)
    a = await d.approvals.decide(current_user(request).id, param_id(id, "id"),
                                 req.expected_revision, req.decision)
    return data(200, a)
